src/pyhf/contrib/viz/brazil.py

```diff
@@ -24,27 +91,213 @@ def plot_results(ax, mutests, tests, test_size=0.05):
         mutests (:obj:`list` or :obj:`array`): The values of the POI where the
           hypothesis tests were performed.
-        tests (:obj:`list` or :obj:`array`): The :math:`\\mathrm{CL}_{s}` values
+        tests (:obj:`list` or :obj:`array`): The :math:`\mathrm{CL}_{s}` values
           from the hypothesis tests.
-        test_size (:obj:`float`): The size, :math:`\\alpha`, of the test.
+        test_size (:obj:`float`): The size, :math:`\alpha`, of the test.
+
+    Returns:
+        :obj:`~pyhf.contrib.viz.brazil.BrazilBandContainer`: A container of :obj:`matplotlib.artist` drawn.
     """
     cls_obs = np.array([test[0] for test in tests]).flatten()
     cls_exp = [np.array([test[1][i] for test in tests]).flatten() for i in range(5)]
-    ax.plot(mutests, cls_obs, c='black')
-    for idx, color in zip(range(5), 5 * ['black']):
-        ax.plot(
-            mutests, cls_exp[idx], c=color, linestyle='dotted' if idx != 2 else 'dashed'
+
+    line_color = kwargs.pop("color", "black")
+    (cls_obs_line,) = ax.plot(
+        mutests, cls_obs, color=line_color, label=r"$\mathrm{CL}_{s}$"
+    )
+
+    cls_exp_lines = []
+    for idx, color in zip(range(5), 5 * [line_color]):
+        (_cls_exp_line,) = ax.plot(
+            mutests,
+            cls_exp[idx],
+            color=color,
+            linestyle="dotted" if idx != 2 else "dashed",
+            label=None if idx != 2 else r"$\mathrm{CL}_{s,\mathrm{exp}}$",
         )
-    ax.fill_between(mutests, cls_exp[0], cls_exp[-1], facecolor='yellow')
-    ax.fill_between(mutests, cls_exp[1], cls_exp[-2], facecolor='green')
-    ax.plot(mutests, [test_size] * len(mutests), c='red')
+        cls_exp_lines.append(_cls_exp_line)
+    one_sigma_band = ax.fill_between(
+        mutests,
+        cls_exp[0],
+        cls_exp[-1],
+        facecolor="yellow",
+        label=r"$\pm2\sigma$ $\mathrm{CL}_{s,\mathrm{exp}}$",
+    )
+    two_sigma_band = ax.fill_between(
+        mutests,
+        cls_exp[1],
+        cls_exp[-2],
+        facecolor="green",
+        label=r"$\pm1\sigma$ $\mathrm{CL}_{s,\mathrm{exp}}$",
+    )
+
+    test_size_color = kwargs.pop("test_size_color", "red")
+    test_size_linestyle = kwargs.pop("test_size_linestyle", "solid")
+    (test_size_line,) = ax.plot(
+        mutests,
+        [test_size] * len(mutests),
+        color=test_size_color,
+        linestyle=test_size_linestyle,
+        label=rf"$\alpha={test_size}$",
+    )
     ax.set_ylim(0, 1)
 
-    ax.set_xlabel(r"$\mu$ (POI)")
-    ax.set_ylabel(r"$\mathrm{CL}_{s}$")
+    x_label = kwargs.pop("xlabel", r"$\mu$ (POI)")
+    y_label = kwargs.pop("ylabel", r"$\mathrm{CL}_{s}$")
+    ax.set_xlabel(x_label)
+    ax.set_ylabel(y_label)
+
+    return BrazilBandContainer(
+        (
+            cls_obs_line,
+            cls_exp_lines,
+            one_sigma_band,
+            two_sigma_band,
+            test_size_line,
+        )
+    )
+
+
+def plot_cls_components(
```

Comment:
any reason we could't combine this witth `plot_results` (i.e. just as kwargs whether or not to add the component arttists?)

Reply:
@lukasheinrich @kratsg I think this comes down to what we want the API to be like. For `plot_cls_components` as it is now it requires that the `results` returned by `hypotest` had the `return_expected_set=True, return_tail_probs=True` kwargs so that the CL_s+b and CL_b components can be drawn. If we integrate the `plot_cls_components` functionality into `plot_results` via kwargs then we need to make it clear to users that the kwargs they select also affect the arguments that they're providing.

I can see the benefits of separating out APIs that have different argument requirements and have the kwargs only affect those specific arguments. Though I do understand what you're saying RE: the convenience of not having to change your API to add more context to your visualization, but rather just throw on some kwargs.

I'll take some time today/tonight to play with this a bit and see what the revision looks like.

Reply:
@lukasheinrich I think I'm convinced of your idea as adopting this would also make `plot_results` more robust as it could handle more information in the `tests` argument to `plot_results`.